src/mpcCoreKit.ts

```diff
@@ -851,13 +853,23 @@ export class Web3AuthMPCCoreKit implements ICoreKit {
   }
 
   protected async atomicSync<T>(f: () => Promise<T>): Promise<T> {
+    const isFirstAtomicCall = !this.atomicInProgress;
+    this.atomicInProgress = true;
+
     this.tkey.manualSync = true;
     try {
       const r = await f();
-      await this.commitChanges();
+      if (isFirstAtomicCall && this.atomicInProgress) {
+        if (!this.options.manualSync) {
+          await this.commitChanges();
+        }
+      }
       return r;
     } finally {
-      this.tkey.manualSync = this.options.manualSync;
+      if (isFirstAtomicCall) {
+        this.atomicInProgress = false;
+        this.tkey.manualSync = this.options.manualSync;
+      }
```

Comment:
this doesn't work. there is no guarantee that the first atomic call finishes last.
so you may set `atomicInProgress = false` here even if an atomic call is still in progress.

if you want to make sure that you track the state of multiple async calls correctly, then you probably need to use something like a counter.